### DISCUSSION TITLE
TextLayer Unicode characterSet

### DISCUSSION THREAD

**Discussion by jfuehner** [visgl/deck.gl#6302] [Q&A] [answered]
Having some difficulties getting the expected unicode characters to display when a custom `characterSet` is supplied to the `TextLayer`

For example, take the below array of codePoints

```
const characterCodepoints = [1583,1585,1740,1575,1574,1746,32,1587,1606,1583,1726]
console.log('characterCodepoints fromCodePoint', String.fromCodePoint(...characterCodepoints));
```

The console output is:

 دریائے سندھ

Which [translates](https://translate.google.com/?sl=auto&tl=en&text=%D8%AF%D8%B1%DB%8C%D8%A7%D8%A6%DB%92%20%D8%B3%D9%86%D8%AF%DA%BE&op=translate) to **Indus River** which is correct.

However, I can’t figure out how to convert the Unicode string “ دریائے سندھ" into the appropriate characters to supply to the `characterSet` prop.

Doing something like…

Array.from(String.fromCodePoint(...characterCodepoints))

yields incorrect Urdu characters that do not match up. You can see the output in this [CodePen](https://codepen.io/jfuehner/pen/NWvGwqW?editors=1111)

**Comment by Pessimistress** [comment, marked as answer]
TextLayer does not support right-to-left languages at the moment. You may use the IconLayer to render these.

**Reply by jfuehner** [reply]
@Pessimistress would you be able to provide a bit more information or direct me to an example on how I could use the `IconLayer` in this manner? Would I put the unicode strings into the `iconAtlas`?

Should I open a FR for RTL support for the `TextLayer`?

**Reply by Pessimistress** [reply]
Text rendering is very complicated. TextLayer currently relies on the assumption that the same character always look the same wherever they are in a string. This is not true for RTL languages.

Use IconLayer: https://deck.gl/docs/developer-guide/loading-data#example-use-image-from-a-programmatically-generated-svg-string

**Reply by jfuehner** [reply]
@Pessimistress thanks for the assistance. Totally understand Text rendering is complicated. 

To provide some additional use-case context we are displaying some layer data pulled in from OSM sources which unfortunately has RTL-based text mixed with LTR text. 

With our TextLayer implementation we are using the binary based attributes to keep as much as possible on the GPU. 

If we switch these layers to IconLayer would you anticipate we see much of a reduction in performance since with the IconLayer we cannot use a binary approach and things will be CPU-bound?

**Reply by jfuehner** [reply]
@Pessimistress as a follow up, it looks like MapBox GL JS had similar RTL text rendering issues back in 2017 which they solved.

https://blog.mapbox.com/improving-arabic-and-hebrew-text-in-map-labels-fd184cf5ebd1

Perhaps using their https://github.com/mapbox/mapbox-gl-rtl-text library would allow for rendering the appropriate glyphs via the TextLayer?

**Reply by jfuehner** [reply]
To close the loop on this for others. I was able to use the [mapbox-gl-rtl-text](https://github.com/mapbox/mapbox-gl-rtl-text) library to convert my RTL text characters into the appropriate order. 

With the correct character order I was able to still use the deck TextLayer as opposed to the IconLayer.

**Reply by Pessimistress** [reply]
That's awesome. Would you consider adding your solution to https://deck.gl/docs/api-reference/layers/text-layer#remarks ?

**Reply by frandieguez** [reply]
Hi @jfuehner,

would you mind sharing a code  example using the mapbox-gl-rtl-text? I'm struggling to use it cause my map keeps rendering the text in reverse order.

I'm using the next snippet
```
new TextLayer(
        this.getSubLayerProps({
          ...commonTextProps,
          id: `text-layer`,
          data: data,
          fontWeight: 600,
          getAlignmentBaseline: (d: any) => "top",
          getPixelOffset: [0, 5],
          getPosition: (d: any) => d.geometry.coordinates,
          getSize: 30,
          getText: (d: any) => {
            let text = getTextInLang(d.properties, this.props.lang);

            text = rtlText.applyArabicShaping(text);
            text = rtlText.processBidirectionalText(text, []);
            return text.join("\n");
          },
          getTextAnchor: "middle",
          outlineColor: [255, 255, 255, 200],
          outlineWidth: 0.6,
        })
      ),
```

And in the screenshot below you can see that on the map the text is rendered wrong compared with the same text rendered in a div element by the browsser.
 
![Captura de pantalla de 2022-12-07 12-09-37](https://user-images.githubusercontent.com/4584/206164524-2140822f-8cc7-4044-8a9b-229e4936f651.png)
